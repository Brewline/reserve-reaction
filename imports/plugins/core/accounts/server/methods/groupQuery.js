import { Meteor } from "meteor/meteor";

/**
 * @name groupQuery
 * @method
 * @memberof Accounts/NoMeteorQueries
 * @summary query the Groups collection and return group data
 * @param {Object} context - an object containing the per-request state
 * @param {String} id - id of group to query
 * @return {Object} group object
 */
export async function groupQuery(context, id) {
  const { collections, userHasPermission, userId } = context;
  const { Accounts, Groups } = collections;

  const group = await Groups.findOne({ _id: id });
  if (!group) throw new Meteor.Error("not-found", "There is no group with this ID");

  // If the user has sufficient permissions, then allow them to find any group by ID
  if (userHasPermission(["owner", "admin", "reaction-accounts"], group.shopId)) return group;

  // Otherwise, only let users see groups that they are members of
  const userAccount = await Accounts.findOne({
    _id: userId,
    groups: id
  }, {
    projection: {
      _id: 1
    }
  });

  // If user is not found, throw an error
  if (!userAccount) throw new Meteor.Error("access-denied", "User does not have permissions to view groups");

  return group;
}

/**
 * @name groupsQuery
 * @method
 * @memberof Accounts/NoMeteorQueries
 * @summary query the Groups collection and return a MongoDB cursor
 * @param {Object} context - an object containing the per-request state
 * @param {String} shopId - shop ID to get groups for
 * @return {Object} Groups collection cursor
 */
export async function groupsQuery(context, shopId) {
  const { collections, userHasPermission, userId } = context;
  const { Accounts, Groups } = collections;

  if (userHasPermission(["owner", "admin", "reaction-accounts"], shopId)) {
    // find groups by shop ID
    return Groups.find({ shopId });
  }

  const userAccount = Accounts.findOne({
    _id: userId
  }, {
    projection: {
      groups: 1
    }
  });

  // If user is not found, throw an error
  if (!userAccount) throw new Meteor.Error("access-denied", "User does not have permissions to view groups");

  // find groups by shop ID limited to those the current user is in
  return Groups.find({
    _id: {
      $in: userAccount.groups || []
    },
    shopId
  });
}
