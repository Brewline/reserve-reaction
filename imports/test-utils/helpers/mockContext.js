const mockContext = {
  collections: {},
  shopId: "FAKE_SHOP_ID",
  userHasPermission: jest.fn().mockName("userHasPermission"),
  userId: "FAKE_USER_ID"
};

[
  "Accounts",
  "Assets",
  "Cart",
  "Catalog",
  "Emails",
  "Groups",
  "Inventory",
  "Logs",
  "MediaRecords",
  "Notifications",
  "Orders",
  "Packages",
  "Products",
  "Revisions",
  "roles",
  "SellerShops",
  "Shipping",
  "Shops",
  "Sms",
  "Tags",
  "Templates",
  "Themes",
  "Translations",
  "users"
].forEach((collectionName) => {
  mockContext.collections[collectionName] = {
    insert() {
      throw new Error("insert mongo method is deprecated, use insertOne or insertMany");
    },
    remove() {
      throw new Error("remove mongo method is deprecated, use deleteOne or deleteMany");
    },
    update() {
      throw new Error("update mongo method is deprecated, use updateOne or updateMany");
    },
    deleteOne: jest.fn().mockName(`${collectionName}.deleteOne`),
    deleteMany: jest.fn().mockName(`${collectionName}.deleteMany`),
    find: jest
      .fn()
      .mockName(`${collectionName}.find`)
      .mockReturnThis(),
    findOne: jest.fn().mockName(`${collectionName}.findOne`),
    insertOne: jest.fn().mockName(`${collectionName}.insertOne`),
    insertMany: jest.fn().mockName(`${collectionName}.insertMany`),
    toArray: jest.fn().mockName(`${collectionName}.toArray`),
    updateOne: jest.fn().mockName(`${collectionName}.updateOne`),
    updateMany: jest.fn().mockName(`${collectionName}.updateMany`)
  };
});

mockContext.collections.Media = {
  find: jest.fn().mockName("Media.find"),
  findLocal: jest.fn().mockName("Media.findLocal"),
  findOne: jest.fn().mockName("Media.findOne"),
  findOneLocal: jest.fn().mockName("Media.findOneLocal"),
  insert: jest.fn().mockName("Media.insert"),
  update: jest.fn().mockName("Media.update"),
  remove: jest.fn().mockName("Media.remove")
};

export default mockContext;
