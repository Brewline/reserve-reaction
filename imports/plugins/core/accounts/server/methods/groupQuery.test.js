import mockContext from "/imports/test-utils/helpers/mockContext";
import { groupQuery, groupsQuery } from "./groupQuery";

const fakeShopId = "FAKE_SHOP_ID";
const fakeGroup = { _id: "FAKE_GROUP_ID", name: "fake", shopId: fakeShopId };
const fakeAccount = { _id: "FAKE_ACCOUNT_ID", groups: ["group1", "group2"] };

beforeEach(() => {
  jest.resetAllMocks();
});

describe("groupQuery", () => {
  test("throws not-found if the group does not exist", async () => {
    mockContext.collections.Groups.findOne.mockReturnValueOnce(undefined);
    const result = groupQuery(mockContext, fakeGroup._id);
    return expect(result).rejects.toThrowErrorMatchingSnapshot();
  });

  test("returns the group if userHasPermission returns true", async () => {
    mockContext.collections.Groups.findOne.mockReturnValueOnce(fakeGroup);
    mockContext.userHasPermission.mockReturnValueOnce(true);
    const result = await groupQuery(mockContext, fakeGroup._id);
    expect(mockContext.collections.Groups.findOne).toHaveBeenCalledWith({ _id: fakeGroup._id });
    expect(mockContext.userHasPermission).toHaveBeenCalledWith(["owner", "admin", "reaction-accounts"], fakeGroup.shopId);
    expect(result).toEqual(fakeGroup);
  });

  test("returns the group if userHasPermission returns false but the current user is in the group", async () => {
    mockContext.collections.Groups.findOne.mockReturnValueOnce(fakeGroup);
    mockContext.userHasPermission.mockReturnValueOnce(false);
    mockContext.collections.Accounts.findOne.mockReturnValueOnce(fakeAccount);
    const result = await groupQuery(mockContext, fakeGroup._id);
    expect(mockContext.collections.Groups.findOne).toHaveBeenCalledWith({ _id: fakeGroup._id });
    expect(mockContext.userHasPermission).toHaveBeenCalledWith(["owner", "admin", "reaction-accounts"], fakeGroup.shopId);
    expect(mockContext.collections.Accounts.findOne).toHaveBeenCalledWith({
      _id: mockContext.userId,
      groups: fakeGroup._id
    }, {
      projection: {
        _id: 1
      }
    });
    expect(result).toEqual(fakeGroup);
  });

  test("throws access-denied if not allowed", async () => {
    mockContext.collections.Groups.findOne.mockReturnValueOnce(fakeGroup);
    mockContext.userHasPermission.mockReturnValueOnce(false);
    mockContext.collections.Accounts.findOne.mockReturnValueOnce(undefined);
    const result = groupQuery(mockContext, fakeGroup._id);
    return expect(result).rejects.toThrowErrorMatchingSnapshot();
  });
});

describe("groupsQuery", () => {
  test("returns the groups cursor if userHasPermission returns true", async () => {
    mockContext.collections.Groups.find.mockReturnValueOnce("CURSOR");
    mockContext.userHasPermission.mockReturnValueOnce(true);
    const result = await groupsQuery(mockContext, fakeShopId);
    expect(mockContext.collections.Groups.find).toHaveBeenCalledWith({ shopId: fakeShopId });
    expect(mockContext.userHasPermission).toHaveBeenCalledWith(["owner", "admin", "reaction-accounts"], fakeShopId);
    expect(result).toBe("CURSOR");
  });

  test("returns the groups cursor for groups the current user is in, if userHasPermission returns false", async () => {
    mockContext.collections.Groups.find.mockReturnValueOnce("CURSOR");
    mockContext.userHasPermission.mockReturnValueOnce(false);
    mockContext.collections.Accounts.findOne.mockReturnValueOnce(fakeAccount);
    const result = await groupsQuery(mockContext, fakeShopId);
    expect(mockContext.collections.Groups.find).toHaveBeenCalledWith({
      _id: { $in: fakeAccount.groups },
      shopId: fakeShopId
    });
    expect(mockContext.userHasPermission).toHaveBeenCalledWith(["owner", "admin", "reaction-accounts"], fakeShopId);
    expect(result).toBe("CURSOR");
  });

  test("throws access-denied if not allowed", async () => {
    mockContext.userHasPermission.mockReturnValueOnce(false);
    mockContext.collections.Accounts.findOne.mockReturnValueOnce(undefined);
    const result = groupsQuery(mockContext, fakeShopId);
    return expect(result).rejects.toThrowErrorMatchingSnapshot();
  });
});
