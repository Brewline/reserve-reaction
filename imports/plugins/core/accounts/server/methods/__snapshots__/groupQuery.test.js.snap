// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`groupQuery throws access-denied if not allowed 1`] = `"access-denied"`;

exports[`groupQuery throws not-found if the group does not exist 1`] = `"not-found"`;

exports[`groupsQuery throws access-denied if not allowed 1`] = `"access-denied"`;
